File: clients/typescript-sdk/src/outbox/create-audit-log-dto.ts
import { redactAuditData } from "./audit-redaction.js";

/**
 * DTO for creating an audit log entry in the outbox.
 *
 * Uses an immutable builder pattern - all `with*()` methods return a new instance.
 *
 * `operationData` is redacted (see `redactAuditData`) before it is
 * serialised into the outbox payload, so passwords, tokens and other
 * secret-shaped fields never reach `aud_logs`. Pass a masked-fields list to
 * `withOperationData` for fields the name rule alone would not catch.
 *
 * @example
 * ```typescript
 * const auditLog = CreateAuditLogDto
 *   .create('User', '0HZXEQ5Y8JY5Z', 'CREATE')
 *   .withOperationData({ email: 'user@example.com', name: 'John' })
 *   .withPrincipalId('0HZXEQ5Y8JY5A')
 *   .withSource('user-service');
 * ```
 */
export class CreateAuditLogDto {
	readonly entityType: string;
	readonly entityId: string;
	readonly operation: string;
	readonly operationData: Record<string, unknown> | null;
	readonly maskedFields: readonly string[];
	readonly principalId: string | null;
	readonly performedAt: Date | null;
	readonly source: string | null;
	readonly correlationId: string | null;
	/** FlowCatalyst application (by code) this audit log belongs to; resolved to an application_id at ingest. */
	readonly applicationCode: string | null;
	/** FlowCatalyst client (by code) this audit log belongs to; resolved to a client_id at ingest. */
	readonly clientCode: string | null;
	readonly metadata: Record<string, string>;
	readonly headers: Record<string, string>;

	private constructor(params: {
		entityType: string;
		entityId: string;
		operation: string;
		operationData?: Record<string, unknown> | null;
		maskedFields?: readonly string[];
		principalId?: string | null;
		performedAt?: Date | null;
		source?: string | null;
		correlationId?: string | null;
		applicationCode?: string | null;
		clientCode?: string | null;
		metadata?: Record<string, string>;
		headers?: Record<string, string>;
	}) {
		this.entityType = params.entityType;
		this.entityId = params.entityId;
		this.operation = params.operation;
		this.operationData = params.operationData ?? null;
		this.maskedFields = params.maskedFields ?? [];
		this.principalId = params.principalId ?? null;
		this.performedAt = params.performedAt ?? null;
		this.source = params.source ?? null;
		this.correlationId = params.correlationId ?? null;
		this.applicationCode = params.applicationCode ?? null;
		this.clientCode = params.clientCode ?? null;
		this.metadata = params.metadata ?? {};
		this.headers = params.headers ?? {};
	}

	static create(
		entityType: string,
		entityId: string,
		operation: string,
	): CreateAuditLogDto {
		return new CreateAuditLogDto({ entityType, entityId, operation });
	}

	/**
	 * @param maskedFields Top-level field names to mask in addition to the
	 *   name rule (see `redactAuditData`), e.g. a config value whose secrecy
	 *   depends on a sibling field.
	 */
	withOperationData(
		operationData: Record<string, unknown>,
		maskedFields: readonly string[] = [],
	): CreateAuditLogDto {
		return new CreateAuditLogDto({ ...this.toParams(), operationData, maskedFields });
	}

	withPrincipalId(principalId: string): CreateAuditLogDto {
		return new CreateAuditLogDto({ ...this.toParams(), principalId });
	}

	withPerformedAt(performedAt: Date): CreateAuditLogDto {
		return new CreateAuditLogDto({ ...this.toParams(), performedAt });
	}

	withSource(source: string): CreateAuditLogDto {
		return new CreateAuditLogDto({ ...this.toParams(), source });
	}

	withCorrelationId(correlationId: string): CreateAuditLogDto {
		return new CreateAuditLogDto({ ...this.toParams(), correlationId });
	}

	/** Set the FlowCatalyst application (by code) this audit log belongs to. */
	withApplicationCode(applicationCode: string): CreateAuditLogDto {
		return new CreateAuditLogDto({ ...this.toParams(), applicationCode });
	}

	/** Set the FlowCatalyst client (by code) this audit log belongs to. */
	withClientCode(clientCode: string): CreateAuditLogDto {
		return new CreateAuditLogDto({ ...this.toParams(), clientCode });
	}

	withMetadata(metadata: Record<string, string>): CreateAuditLogDto {
		return new CreateAuditLogDto({
			...this.toParams(),
			metadata: { ...this.metadata, ...metadata },
		});
	}

	withHeaders(headers: Record<string, string>): CreateAuditLogDto {
		return new CreateAuditLogDto({
			...this.toParams(),
			headers: { ...this.headers, ...headers },
		});
	}

	/** Build the audit log payload for the outbox. Filters out null values. */
	toPayload(): Record<string, unknown> {
		return filterNulls({
			entityType: this.entityType,
			entityId: this.entityId,
			operation: this.operation,
			operationData: this.operationData
				? JSON.stringify(redactAuditData(this.operationData, this.maskedFields))
				: null,
			principalId: this.principalId,
			performedAt: (this.performedAt ?? new Date()).toISOString(),
			source: this.source,
			correlationId: this.correlationId,
			applicationCode: this.applicationCode,
			clientCode: this.clientCode,
			metadata: Object.keys(this.metadata).length > 0 ? this.metadata : null,
		});
	}

	private toParams() {
		return {
			entityType: this.entityType,
			entityId: this.entityId,
			operation: this.operation,
			operationData: this.operationData,
			maskedFields: this.maskedFields,
			principalId: this.principalId,
			performedAt: this.performedAt,
			source: this.source,
			correlationId: this.correlationId,
			applicationCode: this.applicationCode,
			clientCode: this.clientCode,
			metadata: this.metadata,
			headers: this.headers,
		};
	}
}

function filterNulls(obj: Record<string, unknown>): Record<string, unknown> {
	const result: Record<string, unknown> = {};
	for (const [key, value] of Object.entries(obj)) {
		if (value !== null && value !== undefined) {
			result[key] = value;
		}
	}
	return result;
}

File: clients/java-sdk/src/main/java/io/flowcatalyst/sdk/outbox/CreateAuditLogDto.java
package io.flowcatalyst.sdk.outbox;

import java.time.Instant;
import java.util.LinkedHashMap;
import java.util.Map;

/**
 * DTO for creating an audit log entry in the outbox. Immutable builder — all
 * {@code with*} methods return a new instance.
 *
 * <pre>{@code
 * var auditLog = CreateAuditLogDto
 *         .create("User", "0HZXEQ5Y8JY5Z", "CREATE")
 *         .withOperationData(Map.of("email", "user@example.com"))
 *         .withPrincipalId("0HZXEQ5Y8JY5A")
 *         .withSource("user-service");
 * }</pre>
 */
public final class CreateAuditLogDto {

    private String entityType;
    private String entityId;
    private String operation;
    private Map<String, Object> operationData;
    private String principalId;
    private Instant performedAt;
    private String source;
    private String correlationId;
    /* FlowCatalyst application (by code); resolved to an application_id at ingest. */
    private String applicationCode;
    /* FlowCatalyst client (by code); resolved to a client_id at ingest. */
    private String clientCode;
    private Map<String, String> metadata = Map.of();
    private Map<String, String> headers = Map.of();

    private CreateAuditLogDto() {}

    private CreateAuditLogDto copy() {
        CreateAuditLogDto c = new CreateAuditLogDto();
        c.entityType = entityType;
        c.entityId = entityId;
        c.operation = operation;
        c.operationData = operationData;
        c.principalId = principalId;
        c.performedAt = performedAt;
        c.source = source;
        c.correlationId = correlationId;
        c.applicationCode = applicationCode;
        c.clientCode = clientCode;
        c.metadata = metadata;
        c.headers = headers;
        return c;
    }

    public static CreateAuditLogDto create(String entityType, String entityId, String operation) {
        CreateAuditLogDto dto = new CreateAuditLogDto();
        dto.entityType = entityType;
        dto.entityId = entityId;
        dto.operation = operation;
        return dto;
    }

    public CreateAuditLogDto withOperationData(Map<String, Object> operationData) {
        CreateAuditLogDto c = copy();
        c.operationData = operationData;
        return c;
    }

    public CreateAuditLogDto withPrincipalId(String principalId) {
        CreateAuditLogDto c = copy();
        c.principalId = principalId;
        return c;
    }

    public CreateAuditLogDto withPerformedAt(Instant performedAt) {
        CreateAuditLogDto c = copy();
        c.performedAt = performedAt;
        return c;
    }

    public CreateAuditLogDto withSource(String source) {
        CreateAuditLogDto c = copy();
        c.source = source;
        return c;
    }

    public CreateAuditLogDto withCorrelationId(String correlationId) {
        CreateAuditLogDto c = copy();
        c.correlationId = correlationId;
        return c;
    }

    /** Set the FlowCatalyst application (by code) this audit log belongs to. */
    public CreateAuditLogDto withApplicationCode(String applicationCode) {
        CreateAuditLogDto c = copy();
        c.applicationCode = applicationCode;
        return c;
    }

    /** Set the FlowCatalyst client (by code) this audit log belongs to. */
    public CreateAuditLogDto withClientCode(String clientCode) {
        CreateAuditLogDto c = copy();
        c.clientCode = clientCode;
        return c;
    }

    /** Merge additional metadata (existing keys are overwritten). */
    public CreateAuditLogDto withMetadata(Map<String, String> extra) {
        CreateAuditLogDto c = copy();
        Map<String, String> merged = new LinkedHashMap<>(metadata);
        merged.putAll(extra);
        c.metadata = Map.copyOf(merged);
        return c;
    }

    /** Merge additional headers (existing keys are overwritten). */
    public CreateAuditLogDto withHeaders(Map<String, String> extra) {
        CreateAuditLogDto c = copy();
        Map<String, String> merged = new LinkedHashMap<>(headers);
        merged.putAll(extra);
        c.headers = Map.copyOf(merged);
        return c;
    }

    public Map<String, String> headers() {
        return headers;
    }

    /**
     * Build the audit log payload for the outbox (nulls omitted).
     * {@code operationData} is embedded as a JSON string; {@code performedAt}
     * defaults to now.
     */
    Map<String, Object> toPayload(com.fasterxml.jackson.databind.ObjectMapper mapper) {
        Map<String, Object> payload = new LinkedHashMap<>();
        payload.put("entityType", entityType);
        payload.put("entityId", entityId);
        payload.put("operation", operation);
        if (operationData != null) {
            try {
                payload.put("operationData", mapper.writeValueAsString(operationData));
            } catch (com.fasterxml.jackson.core.JsonProcessingException e) {
                throw new IllegalArgumentException(
                        "Audit operationData is not serializable to JSON", e);
            }
        }
        putIfNotNull(payload, "principalId", principalId);
        payload.put("performedAt", (performedAt != null ? performedAt : Instant.now()).toString());
        putIfNotNull(payload, "source", source);
        putIfNotNull(payload, "correlationId", correlationId);
        putIfNotNull(payload, "applicationCode", applicationCode);
        putIfNotNull(payload, "clientCode", clientCode);
        if (!metadata.isEmpty()) payload.put("metadata", metadata);
        return payload;
    }

    private static void putIfNotNull(Map<String, Object> map, String key, Object value) {
        if (value != null) map.put(key, value);
    }
}
